### ISSUE TITLE
Poor developer experience for `npm audit fix`

### ISSUE THREAD

**Issue by fregante** [npm/npm#20757]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [x] npm is producing incorrect or undesirable behavior.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

`npm audit fix` is harder to use than it should be.

##### First try

```sh
$ npm audit fix

npm ERR! code EAUDITNOLOCK
npm ERR! audit Neither npm-shrinkwrap.json nor package-lock.json found: Cannot audit a project without a lockfile
npm ERR! audit Try creating one first with: npm i --package-lock-only
```

##### Second try

```sh
$ npm i --package-lock-only
```

Nothing happened because I have `package-lock=false` in my `.npmrc`. So **to use this feature I have to change my config first.**

##### Third try

`npm audit fix` showed me the table and the commands to run, but `npx npm@next audit fix` fixed it automatically, so this part actually works now (or soon)

### supporting information:

 - `npm -v` prints: 6.0.1
 - `node -v` prints: v9.8.0
 - `npm config get registry` prints:https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: macOS 10.13.3
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [x] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by legodude17** [comment]
So, the issue is that `npm audit` needs `package-lock.json`?

**Comment by fregante** [comment]
Yes + the suggested action (`npm i --package-lock-only`) doesn't work until you realize you have to _also_ change your `npmrc`

**Comment by legodude17** [comment]
`npm audit` needing `package-lock.json` is, AFAIK, intended. You could get around changing your `.npmrc` by running `npm i --package-lock-only --package-lock`, IIRC.

**Comment by fregante** [comment]
That extra flag isn’t specified in the suggestion though. 

Either way, could the `audit fix` command take care of itself? I publish npm modules and I don’t need the lock to be there at all.

It’d be nice if I could still use it without having to change my config. It could create the lock on the fly without saving it and requiring any config change.

**Comment by legodude17** [comment]
@bfred-it That sounds like a nice RFC.

**Comment by marijnh** [comment]
At the very least, I feel `npm audit` could spit out some useful information even if there isn't a lockfile. It _can_ apparently scan and find specific problems without lockfile, judging from output like "found 2 vulnerabilities (1 low, 1 critical)". Giving me that information and then refusing to specify what the problem is because I don't have a lockfile doesn't feel helpful.

**Comment by wessberg** [comment]
Given [the announcement](https://blog.npmjs.org/post/175511531085/the-node-security-platform-service-is-shutting) that nsp will shut down by the end of September, I'm about to rely on `npm audit` instead of `nsp check` (which doesn't require a lockfile), and I too find it important that the equivalent npm commands either support security audits without lockfiles (as nsp did) or temporarily generate one for the duration of the audit for those NPM users that has explicitly opted out of lockfiles.
